Update maintain-troubleshoot-database.md (#1102)
MicrosoftDocs/SupportArticles-docs: support/developer/biztalk/management-operations/maintain-troubleshoot-database.md

This article describes how to maintain BizTalk Server databases and how to troubleshoot BizTalk Server database issues.

> ## You must disable the `auto update statistics` and `auto create statistics` options

Reviewer: Don't use code mark in the title.

> You must keep the `auto update statistics` and `auto create statistics` options disabled on the `BizTalkMsgBoxDb` database. To determine whether these settings are disabled, execute the following stored procedures in SQL Server:

Reviewer: No need to change the order.

```sql
EXEC sp_dboption 'BizTalkMsgBoxDB', 'auto create statistics'
EXEC sp_dboption 'BizTalkMsgBoxDB', 'auto update statistics'
```

> You should set the current setting to `OFF`. If this setting is set to `ON`, turn it off by executing the following stored procedures in SQL Server:

Reviewer: ```suggestion
You should set the current setting to `off`. If this setting is set to `on`, turn it off by executing the following stored procedures in SQL Server:
```

```sql

> EXEC sp_dboption 'BizTalkMsgBoxDB', 'auto create statistics', 'off'

Reviewer: For the sp "sp_dboption", it's no longer available in the latest versions of SQL Server, right? Do we need to update the commands by using [ALTER DATABASE SET options](https://learn.microsoft.com/en-us/sql/t-sql/statements/alter-database-transact-sql-set-options?view=sql-server-ver16#auto_create_statistics)?

Reviewer 2: @aartig13, could you help answer this question?

EXEC sp_dboption 'BizTalkMsgBoxDB', 'auto update statistics', 'off'
```

## You must set the Max Degree of Parallelism property correctly

> On the computer that is running SQL Server and hosting the `BizTalkMsgBoxDb` database, set the max degree of parallelism `run_value` and `config_value` properties to a value of **1**. In later SQL versions, it's also possible to specify this setting per database instead of per SQL instance. To determine the Max Degree of Parallelism setting, execute the following stored procedure against the Master database in SQL Server:

Reviewer: **Max Degree of Parallelism**

See https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/configure-the-max-degree-of-parallelism-server-configuration-option?view=sql-server-ver16#SSMSProcedure

Many BizTalk Server tables contain columns that use `DataType` definitions. Online indexing can't be performed in these columns. Therefore, you should never rebuild the BizTalk Server indexes while BizTalk Server processes data.

## Locking, deadlocking, or blocking may occur

Typically, locks and blocks occur in a BizTalk Server environment. However, these locks or blocks don't remain for an extended time. Therefore, blocking, and deadlocking indicate a potential problem.

## You may experience issues with large databases or tables

We have seen that when the `BizTalkMsgBoxDb` database is larger, performance problems can occur. Ideally, the `BizTalkMsgBoxDb` database shouldn't be holding any data. The `BizTalkMsgBoxDb` database should be considered a buffer until the data is processed or moved to the `BizTalkDTADb` or BAM database.

An environment that uses a powerful SQL Server at the back end and many long-running orchestrations may have a `BizTalkMsgBoxDb` database that is larger than 5 GB. A high-volume environment that uses no long-running orchestrations should have a `BizTalkMsgBoxDb` database that is much smaller than 5 GB.

> The BizTalkDTADb database does not have a set size. However, if performance decreases, the database is probably too large. For some customers 20 GB may be considered too large, while 200 GB might work fine with a highly strong SQL server running on multiple CPUs, lots of memory, and a fast network and storage. When you have large BizTalk Server databases, you may experience the following issues:

Reviewer: How about: while for others that with ...., 200G may work fine.

|Table|Description|
|---|---|

> |HostName `Q_Suspended`|This table contains a reference to messages in the Spool table that are associated with suspended instances for the particular host. This table is in the `BizTalkMsgBoxDb` database.|

Reviewer: Not the format that we confirmed offline. If it's a placeholder, use braket for it.

Reviewer 2: I am not sure if it is a placeholder.  I have added the code format just in case. @aartig13, could you please direct us to a person who can confirm this?

The `BizTalkMsgBoxDB` database is a high-volume and high-transaction Online Transaction Processing (OLTP) database. Some deadlocking is expected, and this deadlocking is handled internally by the BizTalk Server engine. When this behavior occurs, no errors are listed in the error logs. When you investigate a deadlock scenario, the deadlock that you are investigating in the output must be correlated with a deadlock error in the event logs.

> The `dequeue` command and some SQL Server Agent jobs are expected to deadlock. Typically, these jobs are selected as deadlock victims. These jobs will appear in a deadlock trace. However, no errors are listed in the event logs. Therefore, this deadlocking is expected, and you can safely ignore the deadlocking with these jobs.

Reviewer: Have you checked the command? We can't use code mark if it's not a command name.

If frequent deadlocks appear in a deadlock trace and if a correlating error is listed in the event logs, you may want the deadlock.

## Step 4: Look for blocked processes

Use Activity Monitor in SQL Server to obtain the server process identifier (SPID) of a locking system process. Then, run SQL Profiler to determine the SQL statement that's executing in the locking SPID.

To troubleshoot a locking and blocking issue in SQL Server, use the PSSDiag for SQL utility to capture all the Transact-SQL events that have the blocking script enabled.

In SQL Server 2005 and later versions, you can specify the **blocked process threshold** setting to determine which SPID or SPIDs are blocking longer than the threshold that you specify.

> For more information about the **blocked process threshold** setting, visit the following MSDN website: [blocked process threshold Server Configuration Option](/sql/database-engine/configure-windows/blocked-process-threshold-server-configuration-option).

Reviewer: ```suggestion
For more information about the **blocked process threshold** setting, see [blocked process threshold Server Configuration Option](/sql/database-engine/configure-windows/blocked-process-threshold-server-configuration-option).
```

To delete all data in the `BizTalkMsgBoxDb` database, use the BizTalk Health Monitor (BHM) tool.

## BizTalkDTADb database purging options

To delete all data from the `BizTalkDTADb` database, use the BizTalk Health Monitor (BHM) tool. Otherwise, use one of the following methods.

> [!NOTE]
> While both methods delete all messages, method 2 is faster.

- Method 1:

  1. Back up all BizTalk Server databases.
  1. Execute the `dtasp_PurgeAllCompletedTrackingData` stored procedure. For more information about the `dtasp_PurgeAllCompletedTrackingData` stored procedure, see [How to Manually Purge Data from the BizTalk Tracking Database](/biztalk/core/how-to-manually-purge-data-from-the-biztalk-tracking-database).

     > [!NOTE]
     > This action deletes all completed messages.

- Method 2:
  1. Back up all BizTalk databases.
  1. Execute the `dtasp_CleanHMData` stored procedure. Use this option only if the `BizTalkDTADb` database contains many incomplete instances that must be removed.

     To do this, follow these steps:

     1. Stop all BizTalk hosts, services, and custom isolated adapters. If you use HTTP or the SOAP adapter, restart the IIS services.

> 

Reviewer: ```suggestion
```

     1. Execute the `dtasp_CleanHMData` stored procedure on the `BizTalkDTADb` database.

> 

Reviewer: ```suggestion
```
